File: DS/LinkedList/Merge2LinkedLists.js
//Merge 2 Linked List

function Node(data, next) {
    this.data = data;
    this.next = next;
}

function LinkedList() {
    this.currNode = null;
    this.smallerNode = null;
}

LinkedList.prototype = {
    add: function(data) {
        var newNode = new Node(data, this.currNode);
        this.currNode = newNode;
        return this.currNode;
    },
    print: function(head) {
        while (head) {
            console.log(head.data);
            head = head.next;
        }
    },
    mergeList: function(head1, head2) {
        if (head1 == null && head2 == null) {
            return null;
        } else if (head1 == null) {
            return head2;
        } else if (head2 == null) {
            return head1;
        }


        if (head1.data > head2.data) {
            head2.next = this.mergeList(head1, head2.next);
            return head2;
        } else {
            head1.next = this.mergeList(head1.next, head2);
            return head1;
        }
    }
}

var li = new LinkedList();
li.add(2);
var head1 = li.add(4);
//li.print(head1);

var li2 = new LinkedList();
li2.add(3);
var head2 = li2.add(1);
//li2.print(head2);

var li3 = new LinkedList();
var merge = li3.mergeList(head1, head2);
li3.print(merge);

File: DS/LinkedList/2/Merge2LinkedLists.py
#Merge two linked list in sorted order

class Node:
    def __init__(self,data,next):
        self.data = data
        self.next = next

class LinkedList:
    def __init__(self):
        self.currNode = None
    
    def add(self,data):
        newNode = Node(data,self.currNode)
        self.currNode = newNode
        return newNode
    
    def printall(self,head):
        while head:
            print head.data
            head = head.next
    
    def merge(self,head1,head2):
        if head1 is None and head2 is None:
            return None
        elif head1 is None:
            return head2
        elif head2 is None:
            return head1
        
        if head1.data > head2.data:
            head2.next = self.merge(head1,head2.next)
            return head2
        else :
            head1.next = self.merge(head1.next, head2)
            return head1         

if __name__ == "__main__":
    li = LinkedList()
    li.add(2)
    head1 = li.add(3)
    #li.printall(head1)

    #print '----------'

    li2 = LinkedList()
    li2.add(4)
    head2 = li2.add(1)
    #li2.printall(head2)

    #print '----------'

    mergeList = li2.merge(head1,head2)
    li.printall(mergeList)
